gym_sts/envs/base.py

```diff
@@ -332,12 +341,20 @@ def step(self, action_id: int) -> Tuple[dict, float, bool, dict]:
             # Maybe check that the new obs is the same as the old one, modulo
             # the error field?
             obs = prev_obs
+            valid_actions = ActionValidators.valid_actions(obs)  # use cached?
         else:
+            valid_actions = ActionValidators.valid_actions(obs)
+            while len(valid_actions) == 0:
```

Comment:
Can we avoid looping forever? It would be nice to raise an exception eventually.

Reply:
Done.